### DISCUSSION TITLE
DBC read to output Live CSV of signals

### DISCUSSION THREAD

**Discussion by mitchdetailed** [cantools/cantools#282] [Ideas]
what cantools functions would / should I use to do the following :

parse a DBC file to extract all Signal Names and units and embed them into a list to write to a file.

when a new message arrives, if it includes signals from the DBC file, write the decoded value to the appropriate CSV column and append the csv line in the file..

The end goal is to use a Live analysis program, megalog viewer HD.

**Comment by andlaus** [comment]
does something like this fit your bill?

```python
#! /usr/bin/env python3

import cantools
import can
import sys

can_device = 'vcan0'
bus = can.interface.Bus(can_device, bustype='socketcan')

db = cantools.database.load_file(sys.argv[1])

while True:
    msg = bus.recv()
    decoded_msg = db.decode_message(msg.arbitration_id, msg.data)

    # do the interesting stuff
    print(f'Received CAN message: {decoded_msg}')
```

Note that this is agnostic of the file format and will work with ARXML, KCD, SYM as well as DBC.

**Comment by mitchdetailed** [comment]
I'll likely have to modify the script a bit, but the expected output would
be as following(let's assume throttle pedal is ID 0X100, 1st byte; and
brake pedal is ID 0X200, 1st byte, 0-100 raw is 0-100%


input messages :
(0.000) can0 100#00
(0.120) can0 200#01
(0.230) can0 100#02
(0.340) can0 200#03


output result :

time,throttle_pedal(%),brake_pedal(%)
0.000,0,
0.120,,1
0.230,2,
0.340,,3

**Reply by andlaus** [reply]
I'd recommend that you don't bother fiddling with the bit values and access the data fields you're interested in like this at the business side of above script:

```python
# [...]

    # do the interesting stuff
    db_msg = db.get_message_by_frame_id(msg.arbitration_id)
    if db_msg.name == 'MyLovelyMessageNameFromTheDBC':
        print(f'{decoded_msg["MyLovelySignalNameFromTheDBC"]}')
```

(add the fields you're intested in.) Assuming that your DBC is correct, there's no need to do low-level stuff like extraction of the signal bits from the raw data and properly scaling it manually. If you're intested in the unit of the signal, you can get this via `db_msg.signals['MyLovelyMessageNameFromTheDBC'].unit` but I suppose that you'll only need this for the headline of your CSV (if at all).

**Comment by tjhowse** [comment]
I have code in a fork that adds a JSON export function to the C code
generated from a DBC. We wanted to decode can to JSON at realtime on a very
busy can bus, and python couldn't keep up with the volume. The C canToJSON
implementation is pretty rough-and-ready, and needs work on numerical value
representation and a performance pass, but it works.

**Reply by andlaus** [reply]
what kind of bus speed and hardware are you using? I haven't checked  this, but I'm pretty sure that above script keeps up with a 500kbps bus on Raspberry pi class hardware...

JSON is not covered by cantools.

**Reply by tjhowse** [reply]
We're running on a relatively old Q7 processor, I'm not sure of the volume of can traffic.

The pure-C can->json code lives here: https://github.com/tjhowse/cantools/blob/master/cantools/database/can/c_source.py#L76

I haven't raised a PR upstream for this as it's not finished, nor tested, nor particularly pretty.

**Reply by andlaus** [reply]
ok, this seems like a pretty specific use case, i.e., not one which is relevant for the general public? Obviously, it is stongly advisable to use generated C source for weak target hardware. For custom generators, I can warmly recommend to use jinja2 as template engine (if cantool's build-in generator is not sufficient for you).

**Comment by mitchdetailed** [comment]
mine is likely a specific use case as well, but here's kinda the principle
I made years ago, entirely manual looping over the CAN message for a match.

https://youtu.be/tNAmKgHUyyw

**Reply by andlaus** [reply]
doesn't the above script work for you? (with slight adaptations, obviously.)

**Comment by mitchdetailed** [comment]
I'll be testing tonight, had some obligations last night

**Comment by mitchdetailed** [comment]
lmk your thoughts, it likely could be cleaned up but is successfully
accomplishing my tasks



On Thu, Feb 25, 2021 at 7:21 AM Mitch Minton <mitchminton@gmail.com> wrote:

> I'll be testing tonight, had some obligations last night
>
> On Thu, Feb 25, 2021, 7:19 AM Andreas Lauser <notifications@github.com>
> wrote:
>
>> doesn't the above script work for you? (with slight adaptations,
>> obviously.)
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/eerimoq/cantools/discussions/282#discussioncomment-405343>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AI62NLTGGXRFDUX7764C2XTTAZE6HANCNFSM4YFACB6A>
>> .
>>
>

-- 
Mitch Minton 615-977-9787
Bridgestone - USW1055
